.github/workflows/maven.yml

```diff
@@ -7,15 +7,31 @@ on: [push, pull_request]
 jobs:
   build:
     runs-on: ubuntu-latest
-
+    strategy:
+      matrix:
+        java: [ '11', '17', '21' ]
+    name: Java ${{ matrix.Java }} CI
     steps:
       - uses: actions/checkout@v4
-      - name: Set up JDK 11
+      - name: Set JDK ${{ matrix.Java }}
         uses: actions/setup-java@v4
         with:
           # Keep this version aligned with pom.xml!
-          java-version: "11"
-          distribution: "temurin"
+          java-version: ${{ matrix.java }}
+          distribution: "temurin" #eclipse distribution
           cache: maven
+      - name: Cache Maven packages
+        uses: actions/cache@v4
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ matrix.Java }}-${{ hashFiles('pom.xml') }}
+          restore-keys: |
+            ${{ runner.os }}-m2-${{ matrix.Java }}
+            ${{ runner.os }}-m2
       - name: Build with Maven
-        run: ./mvnw -B package --file pom.xml
+        run: mvn --batch-mode --update-snapshots package --file pom.xml
+      - run: mkdir staging && cp target/*.jar staging
+      - uses: actions/upload-artifact@v4 #provide nice artifact
```

Comment:
@duttonw can you help me to better understand what this `actions/upload-artifact` is required for / does, here? I've myself so far only used this ([here](https://github.com/vorburger/MariaDB4j/blob/0161b83f15843fd2c4465220d80a4f8fd9c979da/.github/workflows/scorecard.yml#L67)) to upload a SARIF file for an OSSF Scorecard, and its "cousin" `upload-pages-artifact` ([here](https://github.com/enola-dev/enola/blob/c255cacc4279c4a792d16bace7bc64c606198878/.github/workflows/ci.yaml#L140)) for a website. I don't quite understand what this brings here?

Reply:
https://github.com/actions/upload-artifact 
You can see the results here: https://github.com/duttonw/ch.vorburger.exec/actions/runs/8459031016
It surfaces out of the container action anything you tell it to keep for about 90 days.

Reply:
OK